code_generation/rules.py

```diff
@@ -0,0 +1,177 @@
+import logging
+from typing import List, Union
+
+from code_generation.producer import (
+    CollectProducersOutput,
+    Producer,
+    ProducerGroup,
+    TProducerInput,
+    TProducerStore,
+)
+from code_generation.quantity import QuantitiesStore
+
+log = logging.getLogger(__name__)
+
+
+class ProducerRule:
+    def __init__(
+        self,
+        producers: TProducerInput,
+        samples: Union[str, List[str]],
+        scopes: Union[str, List[str]] = "global",
+        invert: bool = False,
+        update_output: bool = True,
+    ):
+        if isinstance(producers, ProducerGroup) or isinstance(producers, Producer):
+            producers = [producers]
+        self.producers = producers
+        if isinstance(samples, str):
+            samples = [samples]
+        self.samples = samples
+        if isinstance(scopes, str):
+            scopes = [scopes]
+        self.scopes = scopes
+        self.invert = invert
+        self.update_output = update_output
+        self.global_scope = "global"
+
+    def set_scopes(self, scopes: List[str]) -> None:
+        if isinstance(scopes, str):
```

Comment:
Considering the type of `scopes`, this should never be true?

Reply:
yes, it is more of a check for users without type annotation checks